### PR TITLE
#164198189 Add support for users to be able to follow other users

### PR DESCRIPTION
**Pull request by devdbrandy**
#### What does this PR do?

Add support for users to be able to follow other users

#### Description of Task to be completed?

- [x] Add endpoint to follow user - `/api/v1/profiles/:username/follow`
- [x] Add endpoint to unfollow user - `/api/v1/users/profiles/:username/unfollow`
- [x] Add endpoint to fetch all user's followers - `/api/v1/users/profiles/followers`
- [x] Add endpoint to fetch users following users - `/api/v1/users/profiles/following`
- [x] Incorporate changes from #21 for improved database with model autoloading
- [x] Refactor models and related codebase to keep a unified code convention

#### How should this be manually tested?

1. Run the following commands on terminal:

```bash
git clone -b feature/164198189/user-followers --single-branch \
https://github.com/andela/apollo-ah-backend.git

cd apollo-ah-backend
cp .env.example .env

# To run test
cp .env.example .env.test

# Install package modules
npm install
```

2. Modify `.env` file with your db credentials

3. Run database migrations with `npm run migrate`

4. Run seeders with `npm run seed`

5. Spin up the server by running `npm run start:dev` on your terminal

6. Test new endpoints listed above ☝️with [Postman](https://www.getpostman.com/downloads/).

❗️ **Important:** `:username` indicate the username to follow

#### Running test locally

Two npm scripts are available to run the test suite:

- `npm run mocha` or `npm run mocha:watch` - The later watches for any file changes and runs the full test suite (without code coverage)
- `npm test` - Performs a single full test suite run, including Istanbul code coverage reporting. Summary coverage reports are written to stdout, and detailed HTML reports are available in /coverage/index.html

**NOTE:** To run test locally, make sure to create `.env.test` file.

#### Any background context you want to provide?

❗️**NOTE:** Database models schema naming has been capitalised across codebase. Ensure that database is cleaned up before running test locally (Contributors only).

#### What are the relevant pivotal tracker stories?

[#164198189](https://www.pivotaltracker.com/story/show/164198189)

#### Screenshots

n/a

#### Questions:

n/a

### CODE REVIEW

File: controllers/followersController.js [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+  static async follow(request, response, next) {
+    const { params: { username } } = request;
+
+    try {
+      const { followable, follower } = await FollowersController.validateFollowable(request);
+      await followable.addFollower(follower);
+
+      return Response.send(response, 200, [], `Successfully followed ${username}`);
+    } catch (error) {
+      console.log(error);
```
[the remark this answers is not in the document]
Comment (devdbrandy) [reply controllers/followersController.js]: Removed

File: test/integrations/routes/profile.test.js [hunk 2]
```diff
@@ -8,6 +8,8 @@ import faker from 'faker';
 // import jwt from 'jsonwebtoken';
 import app from '../../../index';
 import models from '../../../models';
+import { auth } from '../../helpers';
+import { MESSAGE } from '../../../helpers/constants';
```
[the remark this answers is not in the document]
Comment (devdbrandy) [reply test/integrations/routes/profile.test.js]: Remove line

File: controllers/followersController.js [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+   * @memberof FollowersController
+   */
+  static async follow(request, response, next) {
+    const { params: { username } } = request;
+
+    try {
+      const { followable, follower } = await FollowersController.validateFollowable(request);
+      await followable.addFollower(follower);
+
+      return Response.send(response, 200, [], `Successfully followed ${username}`);
```
Comment (aanchirinah) [inline controllers/followersController.js]: #### Nitpick
- Thought we were using constants for statuses?

File: test/integrations/routes/home.test.js [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+          expect(res.body).to.be.an('object');
+          expect(res.body).to.haveOwnProperty('message').to.equal(MESSAGE.WELCOME_MESSAGE);
+          done();
+        });
+    });
+
+    it('should throw an error for invalid route', (done) => {
+      chai
+        .request(app)
+        .get('/404')
```
Comment (aanchirinah) [inline test/integrations/routes/home.test.js]: #### Nitpick

- Seems like a staged test, lol

File: controllers/followersController.js [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+   * @static
+   * @param {object} request - Express Request object
+   * @returns {object} Object holding the information of the followable and follower
+   * @memberof FollowersController
+   */
+  static async validateFollowable(request) {
+    const { user, params: { username } } = request;
+
+    try {
+      const follower = await models.User.findOne({
```
Comment (aanchirinah) [inline controllers/followersController.js]: #### Critical
- Also, check if the user is not deactivated. ie `deletedAt === null`

File: controllers/followersController.js (hunk 3 above)
Comment (devdbrandy) [reply controllers/followersController.js]: Thanks for pointing that, changes have been made.

File: test/integrations/routes/home.test.js (hunk 4 above)
Comment (devdbrandy) [reply test/integrations/routes/home.test.js]: 😄 This was the missing piece. Had to write test against it.

File: controllers/followersController.js (hunk 5 above)
Comment (devdbrandy) [reply controllers/followersController.js]: O yes, that's a good one. Will have that checked also. Thanks

**Comment by devdbrandy** [comment]
Thanks for pointing that @thislekan . Changes have been made.

**Pull request merged**